### ISSUE TITLE
Documentation Calib Regressor

### ISSUE THREAD

**Issue by nilsleh** [awslabs/fortuna#8]
I am applying the [CalibRegressor](https://aws-fortuna.readthedocs.io/en/latest/references/calib_model/calib_model.html#fortuna.calib_model.regression.CalibRegressor) to model outputs and comparing it to a simple implementation from the [A Gentle Intro to Conformal Prediction Paper](https://arxiv.org/abs/2107.07511). I am getting quiet different results and was therefore looking at how the `CalibRegressor` does conformal prediction. I am not sure where to find it in the documentation, and was wondering if you could point me to a paper on which basis the `CalibRegressor` is implemented?

**Comment by gianlucadetommaso** [comment]
Hi Nils,
you can find a usage example of `CalibRegressor` [here](https://aws-fortuna.readthedocs.io/en/latest/usage_modes/model_outputs.html#regression). The [sinusoidal regression](https://github.com/awslabs/fortuna/blob/main/examples/sinusoidal_regression.ipynb) also makes use of it in the section "What if we have model outputs to start from?".

Note that what [CalibRegressor](https://aws-fortuna.readthedocs.io/en/latest/references/calib_model/calib_model.html#fortuna.calib_model.regression.CalibRegressor) does is to take your model outputs, say `f(x)`, and add an additional model on top that acts on the model outputs themselves, say `g(phi, f(x))`. By default, the method will select a form of temperature scaling, but you can use any other deep learning model you want as `g(phi, f(x))`. When you call [CalibRegressor.calibrate](https://aws-fortuna.readthedocs.io/en/latest/references/calib_model/calib_model.html#fortuna.calib_model.regression.CalibRegressor), you are learning the parameters `phi` on a calibration data set.

Now, this is more related with the work from [Guo et al., 2017](https://proceedings.mlr.press/v70/guo17a.html) than with conformal prediction. However, you can take the uncertainty estimates you get from `CalibRegressor` and use conformal prediction from them.

As an example, take the [sinusoidal regression](https://github.com/awslabs/fortuna/blob/main/examples/sinusoidal_regression.ipynb) regression notebook I linked above. As you see, in the very last line some credible intervals on the test set are computed. You can do the very same on a validation data set, and then calibrate these intervals using [QuantileConformalRegressor](https://aws-fortuna.readthedocs.io/en/latest/references/conformal.html#fortuna.conformal.regression.quantile.QuantileConformalRegressor) - see [here](https://aws-fortuna.readthedocs.io/en/latest/usage_modes/uncertainty_estimates.html#id2) for an example.

Please let me know if this clarifies. Happy to help.

**Comment by nilsleh** [comment]
Thanks a lot for your reply, that makes the distinction between the CalibRegressor and ConformalRegressor clear. I made a  [colab notebook](https://colab.research.google.com/drive/1SMt5c3iidTRpUcOxQkWjrnB8GMBrRJGl?usp=sharing) demonstrating the example I am working on. So it seems that QuantleConformalRegressor is the same implementation as [here](https://github.com/aangelopoulos/conformal-prediction/blob/main/notebooks/meps-cqr.ipynb), which uniformly adjusts the quantiles. However, I was expecting that it would be a locally adaptive implementation, because while marginal coverage is achieved, I would expect better empirical conditional coverage even though conditional coverage is not guaranteed by conformal prediction.

**Comment by gianlucadetommaso** [comment]
Hey Nils,
great to see that you could play with this pretty quickly!

I had a look at your notebook. I have a couple of comments.

1) Make sure to calibrate your model outputs on a *validation* set and not on directly on the test set. The validation set should be a holdout set that you have not used for training, and that is not the validation set. In fact, at calibration time you should not have `y_test` available. 

2) Make sure that your calibration has converged. You can do `plt.plot(calib_status["loss"])` and you will see that your loss function is actually decreasing but it seems it hasn't finished converging. You could fix this by configuring your optimizer, for example as follows:

```
from fortuna.calib_model.calib_config import CalibConfig, CalibOptimizer
calib_status = calib_model.calibrate(
    calib_outputs=[your calibration outputs here], 
    calib_targets=[your calibration labels here],
    calib_config=CalibConfig(optimizer=CalibOptimizer(n_epochs=1000))
)
```

3) Just for your information, you can also compute the empirical coverage directly using one of Fortuna's metrics. It gives you the same result.

```
from fortuna.metric.regression import prediction_interval_coverage_probability
picp = prediction_interval_coverage_probability(*test_cred_intervals.T, y_test.cpu().numpy())
```



Perhaps you could try working with a validation set, calibrating your model longer and then let me know what happens? Thanks a lot!

**Comment by nilsleh** [comment]
Thank you for your reply. Regarding your first point my "test" set was more of a calibration set because it is a toy example for evaluation. However, I updated the [notebook](https://colab.research.google.com/drive/1SMt5c3iidTRpUcOxQkWjrnB8GMBrRJGl?usp=sharing) to make it more explicit to include a validation set. I think my main question is why the conformal prediction is not implemented to be adaptive?

**Comment by gianlucadetommaso** [comment]
Nice observation. Fortuna currently implements the conformalized quantile regression (CQR) method as introduced in [Romano et al., 2019](https://arxiv.org/pdf/1905.03222.pdf). I think your example shows pretty nicely the limitations of satisfying a marginal coverage rather than a conditional one. Under this perspective you have two options: try other methods from conformal prediction and see if they might work better; try a method from another framework (e.g. Bayesian inference) and see if you can get the behaviour that you might expect (although you give up on the marginal guarantee in CP).

I would be curious to see what happens if instead of CQR you use Jackknife+ (see [here](https://aws-fortuna.readthedocs.io/en/latest/examples/jackknifeplus_regression.html) for an example). Would you get closer to the desired conditional coverage?

If you have a link to some existing work deriving an adaptive version of CQR, please let me know. We might consider implementing it, if it works better.

**Comment by gianlucadetommaso** [comment]
Hi Nils,
closing this as it's been open for a little while. Feel free to re-open it if there is any follow-up!

**Comment by gianlucadetommaso** [comment]
Hi Nils, getting back to you as we now support BatchMVP, a conformal prediction method that guarantees coverage on sub-groups of the input space. This should considerably improve the results in your notebook example, where your were correctly observing that just satisfying marginal coverage may lead to undercovering on part of the input space, and overcovering on the other.

Please find [here](https://aws-fortuna.readthedocs.io/en/latest/examples/multivalid_coverage.html) an example similar to yours. Let me know if it helps!

**Comment by nilsleh** [comment]
Hi Gianluca, thanks a lot for getting back to me. I am trying it out at the moment but running into google colab installation issues with fortuna again :/ The fix suggested in #7 unfortunately doesn't work anymore, and I am again not sure how to fix it. The google colab notebook is updated [here](https://colab.research.google.com/drive/1SMt5c3iidTRpUcOxQkWjrnB8GMBrRJGl?usp=sharing). Or should I repopen the other issue again?

**Comment by gianlucadetommaso** [comment]
Mh, that's a pain. It is all due to the pre-installed package versions in Colab. Feel free to open the issue again, I'll make time to look into it.

**Comment by gianlucadetommaso** [comment]
Actually, I just looked and it seems I might have found a fix. Please try and run

```
%pip install --upgrade cvxpy numba jedi setuptools aws-fortuna
```

instead of the installation suggested above. It works for me. Let me know!
